src/tool/subcommands/api_cmd/api_run_tests.rs

```diff
@@ -0,0 +1,593 @@
+    where
+        M: RpcMethod<ARITY>,
+    {
+        self.used_methods.push(M::NAME);
+        if let Some(alias) = M::NAME_ALIAS {
+            self.used_methods.push(alias);
+        }
+        self
+    }
+}
+
+pub(super) async fn run_tests(
+    tests: impl IntoIterator<Item = RpcTestScenario> + Clone,
+    client: impl Into<Arc<rpc::Client>>,
+    filter: String,
+) -> anyhow::Result<()> {
+    let client: Arc<rpc::Client> = client.into();
+    if let Some(token) = client.token() {
+        println!("token: {token}");
+    }
+
+    let mut passed = 0;
+    let mut failed = 0;
+    let ignored = 0;
+    let mut filtered = 0;
+
+    println!("running {} tests", tests.clone().into_iter().count());
+
+    for (i, test) in tests.into_iter().enumerate() {
+        if !filter.is_empty() && !test.used_methods.iter().any(|m| m.starts_with(&filter)) {
+            filtered += 1;
+            continue;
+        }
+
+        print!(
+            "test {} ... ",
+            if let Some(name) = test.name {
+                name.to_string()
+            } else {
+                format!("#{i}")
+            }
+        );
+
+        io::stdout().flush()?;
+
+        let result = (test.run)(client.clone()).await;
+
+        match result {
+            Ok(_) => {
+                println!("ok");
+                passed += 1;
+            }
+            Err(e) => {
+                if let Some(expected_msg) = test.should_fail_with {
+                    let err_str = format!("{e:#}");
+                    if err_str.contains(expected_msg) {
+                        println!("ok");
+                        passed += 1;
+                    } else {
+                        println!("FAILED ({e:#})");
+                        failed += 1;
+                    }
+                } else {
+                    println!("FAILED {e:#}");
+                    failed += 1;
+                }
+            }
+        }
+    }
+    let status = if failed == 0 { "ok" } else { "FAILED" };
+    println!(
+        "test result: {status}. {passed} passed; {failed} failed; {ignored} ignored; {filtered} filtered out"
+    );
+    Ok(())
+}
+
+async fn next_tipset(client: &rpc::Client) -> anyhow::Result<()> {
+    async fn close_channel(
+        stream: &mut tokio_tungstenite::WebSocketStream<
+            tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>,
+        >,
+        id: &serde_json::Value,
+    ) -> anyhow::Result<()> {
+        let request = json!({
+            "jsonrpc": "2.0",
+            "id": 1,
+            "method": "xrpc.cancel",
+            "params": [id]
+        });
+
+        stream
+            .send(WsMessage::Text(request.to_string().into()))
+            .await
+            .context("failed to send close channel request")?;
+
+        Ok(())
+    }
+
+    let mut url = client.base_url().clone();
+    url.set_scheme("ws")
+        .map_err(|_| anyhow::anyhow!("failed to set scheme"))?;
+    url.set_path("rpc/v0");
```

Comment:
why `v0`? Is it not available in `v1`? `v0` is considered legacy from what I recall.

Reply:
Ah yes, good catch.